File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/ServerACK.java
package org.emulinker.kaillera.controller.v086.protocol;

import java.nio.ByteBuffer;

import org.emulinker.kaillera.controller.messaging.*;
import org.emulinker.util.*;

public class ServerACK extends ACK
{
	public static final byte	ID		= 0x05;
	public static final String	DESC	= "Server to Client ACK";

	public ServerACK(int messageNumber) throws MessageFormatException
	{
		super(messageNumber, 0, 1, 2, 3);
	}

	public byte getID()
	{
		return ID;
	}

	public String getDescription()
	{
		return DESC;
	}

	public static ServerACK parse(int messageNumber, ByteBuffer buffer) throws ParseException, MessageFormatException
	{
		if (buffer.remaining() < 17)
			throw new ParseException("Failed byte count validation!");

		byte b = buffer.get();
		
		if (b != 0x00)
			throw new MessageFormatException("Invalid " + DESC + " format: byte 0 = " + EmuUtil.byteToHex(b));

		long val1 = UnsignedUtil.getUnsignedInt(buffer);
		long val2 = UnsignedUtil.getUnsignedInt(buffer);
		long val3 = UnsignedUtil.getUnsignedInt(buffer);
		long val4 = UnsignedUtil.getUnsignedInt(buffer);

		if (val1 != 0 || val2 != 1 || val3 != 2 || val4 != 3)
			throw new MessageFormatException("Invalid " + DESC + " format: bytes do not match acceptable format!");
		
		return new ServerACK(messageNumber);
	}
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/ServerACK.kt
package org.emulinker.kaillera.controller.v086.protocol

import java.nio.ByteBuffer
import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.messaging.ParseException
import org.emulinker.util.EmuUtil
import org.emulinker.util.UnsignedUtil

data class ServerACK
    @Throws(MessageFormatException::class)
    constructor(override val messageNumber: Int) : ACK() {

  override val val1 = 0L
  override val val2 = 1L
  override val val3 = 2L
  override val val4 = 3L

  override val messageId = ID

  init {
    validateMessageNumber(messageNumber)
  }

  companion object {
    const val ID: Byte = 0x05

    @Throws(ParseException::class, MessageFormatException::class)
    fun parse(messageNumber: Int, buffer: ByteBuffer): ServerACK {
      if (buffer.remaining() < 17) throw ParseException("Failed byte count validation!")
      val b = buffer.get()
      if (b.toInt() != 0x00) {
        throw MessageFormatException("byte 0 = " + EmuUtil.byteToHex(b))
      }
      val val1 = UnsignedUtil.getUnsignedInt(buffer)
      val val2 = UnsignedUtil.getUnsignedInt(buffer)
      val val3 = UnsignedUtil.getUnsignedInt(buffer)
      val val4 = UnsignedUtil.getUnsignedInt(buffer)
      if (val1 != 0L || val2 != 1L || val3 != 2L || val4 != 3L)
          throw MessageFormatException(
              "Invalid Server to Client ACK format: bytes do not match acceptable format!")
      return ServerACK(messageNumber)
    }
  }
}
